### ISSUE TITLE
[0.6.4] "Elixir LS has crashed" any time an Elixir file is staged or unstaged for commit

### ISSUE THREAD

**Issue by thepeoplesbourgeois** [elixir-lsp/elixir-ls#481]
The bug doesn't occur in 0.6.2, so I'm reverting to that for now.

### Environment

* **Elixir & Erlang versions (elixir --version):**
  * Erlang/OTP 23 [erts-11.1.7] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:1] [hipe] [dtrace]
  * Elixir 1.11.3 (compiled with Erlang/OTP 23)
* **Operating system:** darwin 20.3.0 (macOS Big Sur 11.2)
* **Editor or IDE name (e.g. Emacs/VSCode):** VSCode / VSCodium (exhibited in both)
* **Editor Plugin/LSP Client name:** jakebecker.elixir-ls

### Troubleshooting

- [x] Restart your editor (which will restart ElixirLS) sometimes fixes issues
- [x] Stop your editor, remove the entire `.elixir_ls` directory, then restart your editor
  * NOTE: This will cause you to have to re-run the dialyzer build for your project

If you're experiencing high CPU usage, it is most likely Dialyzer building the PLTs; after it's done the CPU usage should go back to normal. You could also disable Dialyzer in the settings.

### Logs

```
[error] GenServer ElixirLS.LanguageServer.Server terminating
** (FunctionClauseError) no function clause matching in anonymous fn/2 in ElixirLS.LanguageServer.Server.handle_notification/2
  (language_server 0.6.4) lib/language_server/server.ex:402: anonymous fn(
    %{
      "type" => 2,
      "uri" =>
        "git:/Users/newb/workspace/elixir/empty/lib/empty.ex?%7B%22path%22%3A%22%2FUsers%2Fnewb%2Fworkspace%2Felixir%2Fempty%2Flib%2Fempty.ex%22%2C%22ref%22%3A%22HEAD%22%7D"
    },
    %{
      "file:///Users/newb/workspace/elixir/empty/foo.ex" => %ElixirLS.LanguageServer.SourceFile{
        dirty?: false,
        text: "defmodule Foo do
  end
  ",
        version: 17
      },
      "file:///Users/newb/workspace/elixir/empty/lib/empty.ex" =>
        %ElixirLS.LanguageServer.SourceFile{
          dirty?: false,
          version: 1,
          text: """
          defmodule Empty do
            @moduledoc ""\"
            Documentation for 'Empty'.
            ""\"
  
            @doc ""\"
            Hello world.
  
            ## Examples
  
                iex> Empty.hello()
                :world
  
            ""\"
            def hello do
              :world
            end
          end
          """
        }
    }
  ) in ElixirLS.LanguageServer.Server.handle_notification/2
  (elixir 1.11.3) lib/enum.ex:2193: Enum."-reduce/3-lists^foldl/2-0-"/3
  (language_server 0.6.4) lib/language_server/server.ex:402: ElixirLS.LanguageServer.Server.handle_notification/2
  (language_server 0.6.4) lib/language_server/server.ex:164: ElixirLS.LanguageServer.Server.handle_cast/2
  (stdlib 3.14) gen_server.erl:689: :gen_server.try_dispatch/4
  (stdlib 3.14) gen_server.erl:765: :gen_server.handle_msg/6
  (stdlib 3.14) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
Last message: {:"$gen_cast",
{:receive_packet,
 %{
   "jsonrpc" => "2.0",
   "method" => "workspace/didChangeWatchedFiles",
   "params" => %{
     "changes" => [
       %{
         "type" => 2,
         "uri" =>
           "git:/Users/newb/workspace/elixir/empty/lib/empty.ex?%7B%22path%22%3A%22%2FUsers%2Fnewb%2Fworkspace%2Felixir%2Fempty%2Flib%2Fempty.ex%22%2C%22ref%22%3A%22HEAD%22%7D"
       },
       %{
         "type" => 2,
         "uri" =>
           "git:/Users/newb/workspace/elixir/empty/lib/empty.ex?%7B%22path%22%3A%22%2FUsers%2Fnewb%2Fworkspace%2Felixir%2Fempty%2Flib%2Fempty.ex%22%2C%22ref%22%3A%22%22%7D"
       }
     ]
   }
 }}}
State: %ElixirLS.LanguageServer.Server{
  analysis_ready?: true,
  awaiting_contracts: [],
  build_diagnostics: [],
  build_ref: "#Reference<0.992529980.3137077249.260256>",
  build_running?: false,
  client_capabilities: %{
    "textDocument" => %{
      "codeAction" => %{
        "codeActionLiteralSupport" => %{
          "codeActionKind" => %{
            "valueSet" => [
              "",
              "quickfix",
              "refactor",
              "refactor.extract",
              "refactor.inline",
              "refactor.rewrite",
              "source",
              "source.organizeImports"
            ]
          }
        },
        "dynamicRegistration" => true,
        "isPreferredSupport" => true
      }, 
      "codeLens" => %{"dynamicRegistration" => true},
      "colorProvider" => %{"dynamicRegistration" => true},
      "completion" => %{
        "completionItem" => %{
          "commitCharactersSupport" => true,
          "deprecatedSupport" => true,
          "documentationFormat" => ["markdown", "plaintext"],
          "preselectSupport" => true,
          "snippetSupport" => true,
          "tagSupport" => %{"valueSet" => [1]}
        },
        "completionItemKind" => %{
          "valueSet" => [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25]
        },
        "contextSupport" => true,
        "dynamicRegistration" => true
      },
      "declaration" => %{"dynamicRegistration" => true, "linkSupport" => true},
      "definition" => %{"dynamicRegistration" => true, "linkSupport" => true},
      "documentHighlight" => %{"dynamicRegistration" => true},
      "documentLink" => %{"dynamicRegistration" => true, "tooltipSupport" => true},
      "documentSymbol" => %{
        "dynamicRegistration" => true,
        "hierarchicalDocumentSymbolSupport" => true,
        "symbolKind" => %{
          "valueSet" => [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]
        }
      },
      "foldingRange" => %{
        "dynamicRegistration" => true,
        "lineFoldingOnly" => true,
        "rangeLimit" => 5000
      },
      "formatting" => %{"dynamicRegistration" => true},
      "hover" => %{"contentFormat" => ["markdown", "plaintext"], "dynamicRegistration" => true},
      "implementation" => %{"dynamicRegistration" => true, "linkSupport" => true},
      "onTypeFormatting" => %{"dynamicRegistration" => true},
      "publishDiagnostics" => %{
        "relatedInformation" => true,
        "tagSupport" => %{"valueSet" => [1, 2]},
        "versionSupport" => false
      },
      "rangeFormatting" => %{"dynamicRegistration" => true},
      "references" => %{"dynamicRegistration" => true},
      "rename" => %{"dynamicRegistration" => true, "prepareSupport" => true},
      "selectionRange" => %{"dynamicRegistration" => true},
      "signatureHelp" => %{
        "contextSupport" => true,
        "dynamicRegistration" => true,
        "signatureInformation" => %{
          "documentationFormat" => ["markdown", "plaintext"],
          "parameterInformation" => %{"labelOffsetSupport" => true}
        }
      },
      "synchronization" => %{
        "didSave" => true,
        "dynamicRegistration" => true,
        "willSave" => true,
        "willSaveWaitUntil" => true
      },
      "typeDefinition" => %{"dynamicRegistration" => true, "linkSupport" => true}
    },
    "window" => %{"workDoneProgress" => true},
    "workspace" => %{
      "applyEdit" => true,
      "configuration" => true,
      "didChangeConfiguration" => %{"dynamicRegistration" => true},
      "didChangeWatchedFiles" => %{"dynamicRegistration" => true},
      "executeCommand" => %{"dynamicRegistration" => true},
      "symbol" => %{
        "dynamicRegistration" => true,
        "symbolKind" => %{
          "valueSet" => [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]
        }
      },
      "workspaceEdit" => %{
        "documentChanges" => true,
        "failureHandling" => "textOnlyTransactional",
        "resourceOperations" => ["create", "rename", "delete"]
      },
      "workspaceFolders" => true
    }
  },
  dialyzer_diagnostics: [],
  dialyzer_sup: "#PID<0.121.0>",
  load_all_modules?: false,
  needs_build?: false,
  project_dir: "/Users/newb/workspace/elixir/empty",
  received_shutdown?: false,
  requests: %{},
  root_uri: "file:///Users/newb/workspace/elixir/empty",
  server_instance_id: "NUrkfbIxYYQheib-V-kEs8TO4_lHPOKz",
  settings: %{
    "dialyzerEnabled" => true,
    "dialyzerFormat" => "dialyxir_long",
    "dialyzerWarnOpts" => [],
    "enableTestLenses" => false,
    "fetchDeps" => false,
    "mixEnv" => "test",
    "mixTarget" => "",
    "projectDir" => "",
    "signatureAfterComplete" => true,
    "suggestSpecs" => false,
    "trace" => %{"server" => "off"}
  },
  source_files: %{
    "file:///Users/newb/workspace/elixir/empty/foo.ex" => %ElixirLS.LanguageServer.SourceFile{
      dirty?: false,
      text: "defmodule Foo do
end
",
      version: 17
    },
    "file:///Users/newb/workspace/elixir/empty/lib/empty.ex" =>
      %ElixirLS.LanguageServer.SourceFile{dirty?: false, text: "defmodule Empty do
  @moduledoc \"\"\"
  Documentation for `Empty`.
  \"\"\"

  @doc \"\"\"
  Hello world.

  ## Examples

      iex> Empty.hello()
      :world

  \"\"\"
  def hello do
    :world
  end
end
", version: 1}
  },
  supports_dynamic: true
}
```

~~1.  If using a client other than VS Code, please try VSCode's "ElixirLS: Elixir support and debugger" extension. Does it reproduce your failure?~~

2.  Create a new Mix project with `mix new empty`, then open that project with VS Code and open an Elixir file. Is your issue reproducible on the empty project? If not, please publish a repo on Github that does reproduce it.

Yep.

3.  Check the output log by opening `View > Output` and selecting "ElixirLS" in the dropdown. Please include any output that looks relevant. (If ElixirLS isn't in the dropdown, the server failed to launch.)

See under "Logs"

4.  Check the developer console by opening `Help > Toggle Developer Tools` and include any errors that look relevant.

```js
ElixirLS has crashed. See Output panel.
onDidChangeNotification @ workbench.desktop.main.js:sourcemap:738
```

**Comment by thepeoplesbourgeois** [comment]
I swapped out the error log that came from my real project to the one that ElixirLS produced when the same issue occurred in the project created by `mix new empty`. I figured I should give you the least possible amount of noise to work through, and a skeletal non-app seemed like the least noisy place where I could reproduce the bug.

I also formatted it. Because horizontal scrolling is what you give engineers when you want to punish them, and I love ElixirLS far too much to punish its developers.

Y'all rock 🤘
